### PR TITLE
cpu: add stm32f2xx family support (based on #4497)

### PR DESCRIPTION
**Pull request by vincent-d**
Hi,

This pull request adds support for stm32f2xx family.
It's based on #4497, thanks @DipSwitch for the initial work.
The last commit adds the nucleo-f207 board based on stm32f207.

### CODE REVIEW

**Comment by DipSwitch** [comment]
There is an updated version in de EWM... PR

**Comment by toonst** [comment]
@DipSwitch I had a look at the EWM PR, but it is a bit hard to see which parts are for the cpu and which are for the board. Also it seems to incorporate a lot of merges from other PRs. Can you indicate which parts are missing?

We tested this PR extensively on the nucleo-f207 board (and on our own HW) so I think it is ready for a merge.

**Comment by vincent-d** [comment]
What has to be done to have an implementation of the stm32f2 into RIOT ?
What is your feedback on this PR and what exactly prevents it from being merged ?

**Comment by authmillenon** [comment]
Sorry, a significant portion of the RIOT community that usually looks into this kind of PRs is currently occupied with day job work. This should be finished by tonight.

Optional tip: I'm no expert on porting but usually it helps to split up a PR like this into smaller ones, so people aren't overwhelmed by the sheer number of code lines (34700 is quiet a lot).

**Comment by vincent-d** [comment]
@authmillenon thanks for the reply.

I think the high number of added lines comes mainly from the register header files update.

File: boards/nucleo-f207/include/board.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @brief   LED pin definitions and handlers
+ * @{
+ */
+#undef LED0_PIN
+#undef LED0_MASK
+#undef LED0_ON
+#undef LED0_OFF
+#undef LED0_TOGGLE
+#define LED0_PIN            GPIO_PIN(PORT_B, 0)
```
Comment (kaspar030) [inline boards/nucleo-f207/include/board.h]: Could you please group LEDx_* together?

Comment (toonst) [inline boards/nucleo-f207/include/board.h]: done!

**Comment by toonst** [comment]
Hi guys, can we get an update on this?

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 2]
```diff
@@ -48,6 +50,15 @@ enum {
 };
 
 /**
+ * @brief   Available peripheral buses
+ */
+enum {
+    AHB1,           /**< AHB1 bus */
+    AHB2,           /**< AHB2 bus */
+    AHB3            /**< AHB3 bus */
+};
```
Comment (thomaseichinger) [inline cpu/stm32_common/include/periph_cpu_common.h]: I'd move a change here to a separate PR and use it in all other stm32s too then.

Comment (AurelienGONCE) [inline cpu/stm32_common/include/periph_cpu_common.h]: You highlighted a good point. Actually, the number of AHB is not the same for all stm32 CPUs. This 'enum' should be part of the specific periph_cpu.h.

**Comment by toonst** [comment]
@haukepetersen Any idea when you will have time for reviewing this? We'd like to have it inside the next release.

File: tests/periph_uart/main.c [hunk 3]
```diff
@@ -28,6 +28,7 @@
 #include "msg.h"
 #include "ringbuffer.h"
 #include "periph/uart.h"
+#include "uart_stdio.h"
```
Comment (kaspar030) [inline tests/periph_uart/main.c]: what's the reason for this?

**Comment by kaspar030** [comment]
@toonst @haukepetersen is in paternity leave until September.

@emmanuelsearch Could you find out if anyone at FU has this board available?

**Comment by kaspar030** [comment]
I get the following when compiling examples/default:

```
"make" -C /home/kaspar/src/riot.tmp/sys/shell/commands
dac.c: In function 'dac_set':
dac.c:76:19: error: array subscript is above array bounds [-Werror=array-bounds]
     if (dac_config[line].chan) {
         ~~~~~~~~~~^~~~~~
dac.c: In function 'dac_poweron':
dac.c:91:38: error: array subscript is above array bounds [-Werror=array-bounds]
     DAC->CR |= (1 << (16 * dac_config[line].chan));
                            ~~~~~~~~~~^~~~~~
dac.c: In function 'dac_poweroff':
dac.c:96:39: error: array subscript is above array bounds [-Werror=array-bounds]
     DAC->CR &= ~(1 << (16 * dac_config[line].chan));
                             ~~~~~~~~~~^~~~~~
cc1: all warnings being treated as errors
```

    gcc version 6.1.1 20160526 (Arch Repository)

File: tests/periph_uart/main.c (hunk 3 above)
Comment (toonst) [inline tests/periph_uart/main.c]: see ef0aef3164fe267bbdb17a365518bf7e82a9b19c
This makes sure UART_STDIO_DEV is defined, so the init command fails when trying to use the shell UART as a parameter.
Maybe there's a better way to do this?

**Comment by kaspar030** [comment]
> if anyone at FU has this board available?

If no maintainer has one available, I'd say, once everything compiles, let's merge anyways. Code-wise the PR looks very nice.

**Comment by kYc0o** [comment]
I have the board but I'll have access to them until friday, if the compilation issue can be addressed before Friday I can test.

**Comment by toonst** [comment]
adressed in 378a7a51688e5fb14e03a428b22374b53fbfc570. Thanks for reviewing!

**Comment by toonst** [comment]
I cleaned the history with rebase, should be ready for a merge now.

**Comment by kYc0o** [comment]
The commits seem a bit messy... is it possible to squash them and order them by type? for instance all the uart commits in a single one, same for i2c, spi, etc...

**Comment by toonst** [comment]
@kYc0o addressed. Also rebased onto latest master.

**Comment by kYc0o** [comment]
Thanks @toonst ! But sadly it will go into the next release... But as it's ready to merge it will go fast ;)

**Comment by vincent-d** [comment]
Hi guys, is there any chance this PR could be merged soon ?

**Comment by kaspar030** [comment]
Compiles fine, looks OK. How do we proceed?

**Comment by kaspar030** [comment]
I think we should merge this, if @vincent-d promises to help with basic maintenance of the port.

(merging code that maintainers cannot test is a more general discussion, I'll try to find consensus at devel@)

**Comment by vincent-d** [comment]
@kaspar030 of course I can help with the maintenance of the port :)

**Comment by DipSwitch** [comment]
I could also help when I have more spare time. :)

**Comment by kaspar030** [comment]
> I think we should merge this

Noone seems to disagree. ACK when Murdock is fine.

**Comment by kaspar030** [comment]
Murdock complains about extra whitespaces in the vendor headers. Could you add those files to the doxygen excludes?

**Comment by vincent-d** [comment]
Vendor headers excluded from doxygen

**Comment by kaspar030** [comment]
Well, it's the static tests that complain. But we don't care, so go!
